packages/client-python/src/rocketride/core/runtime/ports.py

```
@@ -0,0 +1,36 @@
+        port = base + offset
+        try:
+            # First check if anything is already listening
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                s.settimeout(0.1)
+                result = s.connect_ex(('127.0.0.1', port))
+                if result == 0:
+                    continue  # Something is listening — skip
+
+            # Then verify we can actually bind
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                s.bind(('0.0.0.0', port))
```

Ruff (0.15.7)

[error] Possible binding to all interfaces

(S104)